src/s390x/zipl.rs

```diff
@@ -0,0 +1,148 @@
+use super::dasd::check;
+use crate::errors::*;
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use walkdir::WalkDir;
+
+/////////////////////////////////////////////////////////////////////////////
+// IBM Z Bootloader Support
+/////////////////////////////////////////////////////////////////////////////
+
+/// Generate zipl config and set dasd as boot device
+///
+/// # Arguments
+/// * `boot` - path to boot partition mountpoint, i.e. smth like /boot
+/// * `dasd` - dasd device, i.e. smth like /dev/dasda
+pub fn install_bootloader<P: AsRef<Path>>(boot: P, dasd: &str) -> Result<()> {
+    eprintln!("Installing_bootloader");
+
+    let bootconf = boot_config(&boot)?;
+    let vmlinuz = kernel(&boot)?;
+    let ramfs = initramfs(&boot)?;
+    let zipl_config = "/tmp/zipl.conf";
+    write_boot_options(&bootconf, zipl_config)
+        .chain_err(|| format!("writing boot options to {}", zipl_config))?;
+
+    let status = Command::new("zipl")
+        .arg("--verbose")
+        .arg("-p")
+        .arg(zipl_config)
+        .arg("-i")
+        .arg(vmlinuz.to_str().unwrap())
+        .arg("-r")
+        .arg(ramfs.to_str().unwrap())
+        .arg("--target")
+        .arg(boot.as_ref().to_str().unwrap())
+        .status()
+        .chain_err(|| "failed to execute zipl")?;
+    check(status).chain_err(|| "installing bootloader")?;
+
+    let status = Command::new("chreipl")
+        .arg(dasd)
+        .status()
+        .chain_err(|| "failed to execute chreipl")?;
+    check(status).chain_err(|| format!("setting {} as boot-device", dasd))?;
+    Ok(())
+}
+
+/// Write boot options. On Z we have to generate config file for zipl
+///
+/// # Arguments
+/// * `boot_config` - path to boot config file
+/// * `zipl_config` - path to zipl config file
+fn write_boot_options<P1, P2>(boot_config: P1, zipl_config: P2) -> Result<()>
+where
+    P1: Sized + AsRef<Path>,
+    P2: Sized + AsRef<Path>,
+{
+    let boot_config = boot_config.as_ref();
+    let zipl_config = zipl_config.as_ref();
+    let mut contents = read_options(&boot_config)?;
+    let pat = "options ";
+    let idx = contents
+        .find(pat)
+        .chain_err(|| format!("cannot find: \'{}\'", pat))?
+        + pat.len();
+
+    let s390_options = read_options("/tmp/s390x_opts")?;
+    contents.insert_str(idx, &s390_options);
+
+    eprintln!("Writing boot options");
+    std::fs::write(&boot_config, &contents)
+        .chain_err(|| format!("writing {}", boot_config.display()))?;
+
+    eprintln!("Generating zipl config");
+    let mut contents = &contents[idx..contents.len()];
+    if let Some(idx) = contents.find('\n') {
+        contents = &contents[0..idx];
```

Comment:
We're already pulling in `regex`, so let's use it rather than manual string manipulation.